logging-appender/src/main/java/org/daisy/pipeline/logging/EventBusAppender.java

```diff
@@ -0,0 +1,99 @@
+package org.daisy.pipeline.logging;
+
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+
+import ch.qos.logback.classic.Level;
+import ch.qos.logback.classic.sift.MDCBasedDiscriminator;
+import ch.qos.logback.classic.spi.ILoggingEvent;
+import ch.qos.logback.core.AppenderBase;
+
+import org.daisy.common.messaging.Message;
+import org.daisy.common.messaging.MessageBuliderFactory;
+import org.daisy.pipeline.event.EventBusProvider;
+
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.FrameworkUtil;
+import org.osgi.framework.InvalidSyntaxException;
+
+/**
+ * Configure like this:
+ *
+ * &lt;appender name="EVENTBUS" class="org.daisy.pipeline.logging.EventBusAppender"&gt;
+ *   &lt;filter class="org.daisy.pipeline.logging.ThresholdFilter"&gt;
+ *     &lt;rootLevel&gt;INFO&lt;/rootLevel&gt;
+ *     &lt;loggerLevels&gt;
+ *       cz.vutbr.web=WARN
+ *       org.daisy.common.xproc.calabash.steps.Message=WARN
+ *       com.xmlcalabash.runtime.XAtomicStep=WARN
+ *     &lt;/loggerLevels&gt;
+ *   &lt;/filter&gt;
+ * &lt;/appender&gt;
+ */
+public class EventBusAppender extends AppenderBase<ILoggingEvent> {
+	
+	private EventBusProvider eventBusProvider;
+	private BundleContext bundleContext;
+	private MessageBuliderFactory messageBuilderFactory;
+	private List<ILoggingEvent> eventBuffer;
+	private MDCBasedDiscriminator discriminator;
+	
+	@Override
+	public void start() {
+		bundleContext = FrameworkUtil.getBundle(this.getClass()).getBundleContext();
```

Comment:
Copied from Slack
> @capitancambio Yes, you're right, and this while we are trying to go in the other direction :). I'm not sure there is a way to avoid it. I couldn't use declarative services because it is a fragment bundle. Maybe the only way is to have several artifacts, one for OSGi, and one for each other dependency injection mechanism you want to use?